cmd/prometheus/main.go

```diff
@@ -282,104 +278,182 @@ func main() {
+		g.Add(
+			func() error {
+				db, err := tsdb.Open(
+					cfg.localStoragePath,
+					log.With(logger, "component", "tsdb"),
+					prometheus.DefaultRegisterer,
+					&cfg.tsdb,
+				)
+				if err != nil {
+					level.Error(logger).Log("msg", "Opening storage failed", "err", err)
+					return nil
+				}
+				localStorage.Set(db)
+				close(dbOpen)
+				<-s4
+				return nil
+			},
+			func(err error) {
+				if err := fanoutStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
+				}
+				close(s4)
+			},
+		)
 
-	errc := make(chan error)
-	go func() { errc <- webHandler.Run(ctx) }()
+		// The notifier is a dependency of the rule manager. It has to be
+		// started before and torn down afterwards.
+		g.Add(
+			func() error {
+				notifier.Run()
+				return nil
+			},
+			func(err error) {
+				notifier.Stop()
+			},
+		)
 
-	<-dbOpen
+		//TODO(krasi) refactor ruleManager.Run() to be blocking to avoid using an extra blocking channel
+		s5 := make(chan struct{})
+		g.Add(
+			func() error {
+				ruleManager.Run()
+				<-s5
+				return nil
+			},
+			func(err error) {
+				ruleManager.Stop()
+				close(s5)
+			},
+		)
+		//TODO(krasi) refactor targetManager.Run() to be blocking to avoid using an extra blocking channel
+		s6 := make(chan struct{})
+		g.Add(
+			func() error {
+				targetManager.Run()
+				<-s6
+				return nil
+			},
+			func(err error) {
+				targetManager.Stop()
+				close(s6)
+			},
+		)
+		g.Add(
+			func() error {
+				if err := webHandler.Run(ctx); err != nil {
+					level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+				}
+				return nil
+			},
+			func(err error) {
+				// Shutting down the query engine before the rule manager will cause pending queries
+				// to be canceled and ensures a quick shutdown of the rule manager.
+				cancelCtx()
+			},
+		)
 
-	if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
-		level.Error(logger).Log("msg", "Error loading config", "err", err)
-		os.Exit(1)
 	}
-
-	defer func() {
-		if err := fanoutStorage.Close(); err != nil {
-			level.Error(logger).Log("msg", "Error stopping storage", "err", err)
-		}
-	}()
-
-	// Wait for reload or termination signals.
-	close(hupReady) // Unblock SIGHUP handler.
-
-	// Set web server to ready.
-	webHandler.Ready()
-	level.Info(logger).Log("msg", "Server is ready to receive requests.")
-
-	term := make(chan os.Signal, 1)
-	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
-	select {
-	case <-term:
-		level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
-	case <-webHandler.Quit():
-		level.Warn(logger).Log("msg", "Received termination request via web service, exiting gracefully...")
-	case err := <-errc:
-		level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+	if err := g.Run(); err != nil {
+		level.Error(logger).Log("msg", "Group interrupt", "err", err)
```

Comment:
As a Prometheus user, without knowing the code and specific implementation here, I would not know what "group" is referring to.

Reply:
you mean this line `level.Error(logger).Log("msg", "Group interrupt", "err", err)` ? 
this is not even used as you can see that all groups return `nil` for the error. 

I am still not sure how to organise the group errors as if we move the errors to be handled by the group package we would loose the extra `"msg"` field.

Open to suggestions.